### ISSUE TITLE
Optimisation at times exceeds p_nom for deferrable loads

### ISSUE THREAD

**Issue by purcell-lab** [davidusb-geek/emhass#324]
**Describe the bug**
At times the optimisation allocates power levels which are above p_nom for deferrable loads.

![EMHASS-–-Home-Assistant (1)](https://github.com/davidusb-geek/emhass/assets/79175134/cbfc094d-74e4-408a-ba80-990f41db1f9c)

**To Reproduce**
![Home-Assistant (1)](https://github.com/davidusb-geek/emhass/assets/79175134/865de44a-baa7-4d09-88e8-d5ff88cacd08)

**Expected behavior**
p_deferrable should not exceed p_nom, so in this case it should not exceed 600 W for the 17:00 timesteps.

**Home Assistant installation type**
 - Home Assistant Supervised

**Your hardware**
- OS: HA OS, Windows, Linux, etc
- Architecture: amd64, armhf, armv7, aarch64

**EMHASS installation type**
 - Add-on

**Additional context**
deferrable4 (my hot water heat pump) has an additional constraint to complete by sunset, which occurs at the 17:00 timestep.

I suspect the optimisation is allocating additional power to deferrable4 at this point, because it is constrained..

**Comment by purcell-lab** [comment]
Another user with the same issue.
https://community.home-assistant.io/t/emhass-an-energy-management-for-home-assistant/338126/2528?u=markpurcell

@GeoDerp , can you assist I think there is a logic error in the stop intervals code that it tries to squeeze in all the power required in the time remaining before the stop interval without consideration of the p_nom restriction.

**Comment by GeoDerp** [comment]
Hey @purcell-lab , Unfortunately im not verse in that area of the EMHASS code.
Out of curiosity, What happens if `'def_start_penalty': []` in runtime ?

**Comment by sladey** [comment]
Maybe related to this issues I am see power allocated to deferrable loads when the def_total_hours runtime param contains zero for that particular def load.

**Comment by purcell-lab** [comment]
Some discussions in discord 
https://discord.com/channels/1317814601466773505/1317982438005411902/1330310776514154537

**Comment by purcell-lab** [comment]
Continuing to be an issue, p_nom is 115200, but with an end_timestep constraint it allocates forecast power levels above p_nom.

![Image](https://github.com/user-attachments/assets/156fede6-09ce-46aa-a5ae-a0f558028d5b)

**Comment by davidusb-geek** [comment]
> Continuing to be an issue, p_nom is 115200, but with an end_timestep constraint it allocates forecast power levels above p_nom.

So you see this problem when using the `end_timestep` contraints, without this the nominal power is never exceeded?

**Comment by purcell-lab** [comment]
This issue only appears with the end_timestamp constraint.

It seems to be it is trying to squeeze in additional power above p_nom in order to complete before the end_timestamp, but oddly you can see from my recent example it allocated double p_nom 23040 W in the 06:00 interval, but then allocated 0 W for the 06:30 interval. Instead it could have allocated 11520 W to both intervals, but it chose not too.

**Comment by RafAustralia** [comment]
@purcell-lab @davidusb-geek @GeoDerp 

Hi all. 

Im gonna add my case/experience here. 
Running Geo’s test version. 

Appears like when emhass has long window to plan ev charging. It will slot it in perfectly. 
As the time passes. Prices change. Emhass starts to readjust and starts charging…. It doesnt go full speed its goes under the limit… eg 16a current allowing 11kw draw but it inly uses 7-9kw. 
Then it adapts and stops trying to look for better price. While its doing that it realises it has run out of time and final timestamp is near…. It tries to pack double the load (impossible to achieve) into one or two slots. Ends up running out if time and doesnt make it. 

Thats just one example. 

Second is pool load. Constant 1400w fixed. 
Needs 7hrs a day. Nothing else. 

But if its window narrows or it thinks weather is bad it will attempt to put variable units in constant load. Eg Ive had couple of hrs if planned 8.8kw load which cannot be achieved. 

Perhaps there needs to be a fail safe that if by certain time eg when window of pool hrs near to equal with required load emhass just needs to make a call and run with it?

Hope that makes sense?

**Comment by RafAustralia** [comment]
@purcell-lab @GeoDerp @davidusb-geek 

Update on deferrable load for ev  - the double up. 

Wild guess. 

I think however the timestamp may respond with 15min intervals - somewhere in emhass a part of the code still responds as 1hr. Cos as soon as mini has opportunity of scheduling at least minimum 1hr window it schedules correctly and starts charging quicker too. 

So I think we changed to timestamps but something is holding it back in scheduling less than 1hr period.

Anyhow. My thoughts.

**Comment by RafAustralia** [comment]
Hi all. New issue related to this. 
I have set battery weights last night (that was a fail as they do not seem to stick) but leaving this issue aside it seems to have caused a havoc in my deferrable loads scheduling where these have exceeded their allowed maximum. 
Eg. Ev is 12000 and pool is 1400 as a deferrable load maximum. 
EMHASS started showing EV as 72000 and pool as selling negative value of -28016. 
Nothing worked to rectify the issue. Restarted EMHASS and still nothing. 
I had to change my EV charging end time stamp by 30min to have all issues correct themselves.  
One note:
Just to be clear. System was optimal and not infeasible ay any time

Any thoughts would be great. 

Pics below

![Image](https://github.com/user-attachments/assets/76094ccb-88c8-4f1a-9dca-370311902c5d)
![Image](https://github.com/user-attachments/assets/bfaf38f0-1540-4d67-aad5-87a8c0c2a3a1)
![Image](https://github.com/user-attachments/assets/038bb5d3-8b46-4fcd-b3bb-49609ec25304)
![Image](https://github.com/user-attachments/assets/6cd24770-2283-438d-a7cd-60760ab32f2a)
![Image](https://github.com/user-attachments/assets/b0d4a1e5-b812-44e7-ba1f-d13d1dda281e)
![Image](https://github.com/user-attachments/assets/a5bf2e93-4d99-444d-bc0a-10701b19728d)
![Image](https://github.com/user-attachments/assets/7482a342-0b0b-4eb6-9145-020703700d29)
![Image](https://github.com/user-attachments/assets/31d34cb3-2183-4eaf-9194-16eaa2b1e6c4)
![Image](https://github.com/user-attachments/assets/c576c9db-1aa9-495a-b354-cadf9c3e05e1)
![Image](https://github.com/user-attachments/assets/21a80152-728a-4cd9-8882-c4935e56193b)

**Comment by purcell-lab** [comment]
## Status update — May 2026

**Findings**

- The thread now has independent corroboration from at least three users (@RafAustralia, @sladey, the community-forum post I linked earlier) plus the original report — so this is a reproducible issue, not a one-off.
- @davidusb-geek confirmed the trigger above: it only appears when `end_timestep` is set. The pattern is that the solver tries to "squeeze in" the remaining required energy before the deadline by allocating power above `p_nom` in late timesteps. My 13 Feb 2025 example shows it allocating **2× p_nom** in the 06:00 interval and 0 W in 06:30, when a flat `p_nom` across both intervals would have been feasible — so the solver isn't even using its degrees of freedom efficiently.
- @RafAustralia (Jun 2025) reported a related-looking but probably distinct case where battery weights produced 6× nominal power for EV (72 kW vs 12 kW) and a negative pool load. That looks more like a stale-state / weight-application bug than the `end_timestep` interaction — worth a separate issue.

**Root cause hypothesis**

In the deferrable-load constraint set, when `end_timestep` is active, the constraint is probably written as a sum-over-window inequality (`Σ P_def[t] * Δt ≥ required_energy`) **without** the per-timestep `P_def[t] ≤ p_nom` constraint also being added for the constrained window. Or, equivalently, the per-timestep bound is being added but with a stale value before `end_timestep` is applied. Either way the fix is to make the per-timestep upper bound a hard constraint that applies independently of the deadline.

**Proposed way forward**

1. **Hard per-timestep bound, always.** Audit the deferrable-load constraint construction in `optimization.py` and ensure `P_deferrable[t] ≤ p_nom` is added unconditionally for every timestep `t` in `[start, end_timestep]`, regardless of whether `end_timestep` is also set. This should be a small, localised change.
2. **Pre-publication bound check (workaround / belt-and-braces).** As I suggested earlier in the thread, add an assertion after solve and before publishing the forecast: if any `P_deferrable[t] > p_nom * (1 + eps)` or `< minimum_power`, log a warning and either clamp the published series or surface an `infeasible_by_bound` flag to HA. This protects users even if the constraint construction has a future regression.
3. **Open separate issue for @RafAustralia's case.** The negative pool load + 6× EV nominal power looks like a different bug class (weights + state, not deadlines). Worth isolating with a minimal reproduction.

Happy to put up a PR for option 1 once @davidusb-geek confirms the constraint topology I'm describing matches the actual code.

**Comment by davidusb-geek** [comment]
It seems like a good idea to implement the "Hard per-timestep bound, always" approach. We need to modify _add_deferrable_load_constraints to ensure that regardless of start/end timesteps, the power at any specific moment never exceeds the nominal capacity.

Also yes with point number 3, the "6x nominal power" issue is likely a separate bug related to how battery weights are applied to the cost function. So it might be a good idea to open a separate issue for that specific case once this `end_timestep` fix is merged
